In [1]:
import numpy as np
import pandas as pd
from awesome_functions import *

/home/hskimim/anaconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/home/hskimim/anaconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/home/hskimim/anaconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/home/hskimim/anaconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/home/hskimim/anaconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/home/hskimim/anaconda3/lib/python3.6/im

In [2]:
train_df = pd.read_csv('train.csv')
test_df = pd.read_csv('test.csv')

In [3]:
train_df.tail()
len(train_df['DepartmentDescription'].value_counts())

68

In [4]:
len(train_df[train_df['DepartmentDescription'] == 'MENS WEAR'])

12250

In [5]:
len(train_df[train_df['DepartmentDescription'] == 'MENSWEAR'])

338

In [6]:
train_df['DepartmentDescription'] = train_df['DepartmentDescription'].replace('MENSWEAR','MENS WEAR')
len(train_df['DepartmentDescription'].value_counts())

67

In [7]:
train_df.isna().sum()

TripType                    0
VisitNumber                 0
Weekday                     0
Upc                      4129
ScanCount                   0
DepartmentDescription    1361
FinelineNumber           4129
dtype: int64

In [8]:
df = train_df[['TripType','VisitNumber','DepartmentDescription','FinelineNumber','Upc']]
df.tail()

TripType  VisitNumber DepartmentDescription  FinelineNumber  \
647049        39       191346          PHARMACY OTC          1118.0   
647050        39       191346          FROZEN FOODS          1752.0   
647051        39       191346               PRODUCE          4170.0   
647052         8       191347                 DAIRY          1512.0   
647053         8       191347     GROCERY DRY GOODS          3600.0   

                 Upc  
647049  3.239000e+10  
647050  7.874205e+09  
647051  4.072000e+03  
647052  4.190008e+09  
647053  3.800060e+09

In [9]:
df2 = df[['DepartmentDescription','FinelineNumber']].groupby(by='DepartmentDescription').count()
df2.tail()

FinelineNumber
DepartmentDescription                
SLEEPWEAR/FOUNDATIONS            2376
SPORTING GOODS                   5087
SWIMWEAR/OUTERWEAR                860
TOYS                             6107
WIRELESS                         1898

In [10]:
print(len(train_df))
print(np.sum(df2.values))
print(len(train_df['FinelineNumber']))

647054
642925
647054


In [11]:
len(train_df['FinelineNumber']) - np.sum(df2.values)

4129

DD가 FL을 포함하고 있는데 FL 데이터의 총 갯수가 642925개다. 그리고 DD에게 포함되고 있는 FL의 데이터 갯수가 647054 개로 그 차이가 4129개로 FL의 비어있는 값과 같다. 즉, 있는 데이터면, 전부 DepartmentDescription 에 포함된다.

In [12]:
df3 = df[['FinelineNumber','Upc']].groupby(by='FinelineNumber').count()
df3.tail()

Upc
FinelineNumber     
9974.0           54
9975.0           28
9991.0            1
9997.0           50
9998.0          411

In [13]:
print(len(train_df))
print(np.sum(df3.values))
print(len(train_df['Upc']))

647054
642925
647054


In [14]:
np.sum(df3.values)

642925

In [15]:
len(train_df['Upc']) - np.sum(df3.values)

4129

In [16]:
df4 = df[['DepartmentDescription','Upc']].groupby(by='DepartmentDescription').count()
df4.tail()

Upc
DepartmentDescription      
SLEEPWEAR/FOUNDATIONS  2376
SPORTING GOODS         5087
SWIMWEAR/OUTERWEAR      860
TOYS                   6107
WIRELESS               1898

In [17]:
print(len(train_df))
print(np.sum(df4.values))
print(len(train_df['Upc']))

647054
642925
647054


In [18]:
len(train_df['Upc']) - np.sum(df4.values)

4129

이러한 방식으로는 판별력이 없기 때문에, VisitNumber pivotting

In [19]:
df5 = df[['TripType','VisitNumber','DepartmentDescription','FinelineNumber','Upc']]
df5.tail()

TripType  VisitNumber DepartmentDescription  FinelineNumber  \
647049        39       191346          PHARMACY OTC          1118.0   
647050        39       191346          FROZEN FOODS          1752.0   
647051        39       191346               PRODUCE          4170.0   
647052         8       191347                 DAIRY          1512.0   
647053         8       191347     GROCERY DRY GOODS          3600.0   

                 Upc  
647049  3.239000e+10  
647050  7.874205e+09  
647051  4.072000e+03  
647052  4.190008e+09  
647053  3.800060e+09

In [20]:
df5[df5['DepartmentDescription'].isna()].tail()

TripType  VisitNumber DepartmentDescription  FinelineNumber  Upc
643137        41       190408                   NaN             NaN  NaN
643991        44       190651                   NaN             NaN  NaN
645990        44       191080                   NaN             NaN  NaN
645991        44       191080                   NaN             NaN  NaN
645992        44       191080                   NaN             NaN  NaN

In [21]:
for i in train_df['TripType'].unique():
     locals()['df_' + str(i)] = train_df[['TripType','DepartmentDescription','FinelineNumber','Upc']][train_df['TripType'] == i] 

In [22]:
testing_df = train_df[['DepartmentDescription','FinelineNumber']]

In [23]:
train_df['DepartmentDescription'].unique()

array(['FINANCIAL SERVICES', 'SHOES', 'PERSONAL CARE',
       'PAINT AND ACCESSORIES', 'DSD GROCERY', 'MEAT - FRESH & FROZEN',
       'DAIRY', 'PETS AND SUPPLIES', 'HOUSEHOLD CHEMICALS/SUPP', nan,
       'IMPULSE MERCHANDISE', 'PRODUCE', 'CANDY, TOBACCO, COOKIES',
       'GROCERY DRY GOODS', 'BOYS WEAR', 'FABRICS AND CRAFTS',
       'JEWELRY AND SUNGLASSES', 'MENS WEAR', 'ACCESSORIES',
       'HOME MANAGEMENT', 'FROZEN FOODS', 'SERVICE DELI',
       'INFANT CONSUMABLE HARDLINES', 'PRE PACKED DELI', 'COOK AND DINE',
       'PHARMACY OTC', 'LADIESWEAR', 'COMM BREAD', 'BAKERY',
       'HOUSEHOLD PAPER GOODS', 'CELEBRATION', 'HARDWARE', 'BEAUTY',
       'AUTOMOTIVE', 'BOOKS AND MAGAZINES', 'SEAFOOD', 'OFFICE SUPPLIES',
       'LAWN AND GARDEN', 'SHEER HOSIERY', 'WIRELESS', 'BEDDING',
       'BATH AND SHOWER', 'HORTICULTURE AND ACCESS', 'HOME DECOR', 'TOYS',
       'INFANT APPAREL', 'LADIES SOCKS', 'PLUS AND MATERNITY',
       'ELECTRONICS', 'GIRLS WEAR, 4-6X  AND 7-14', 'BRAS & SHAPEWEAR',

In [24]:
for i,name in enumerate(train_df['DepartmentDescription'].unique()):
     locals()['df_' + str(i)] = train_df[['DepartmentDescription','FinelineNumber']][train_df['DepartmentDescription'] == name] 

In [25]:
for i,df in enumerate(train_df['DepartmentDescription'].unique()):
    print(df,len(locals()['df_' + str(i)]))

FINANCIAL SERVICES 10723
SHOES 6099
PERSONAL CARE 41969
PAINT AND ACCESSORIES 1720
DSD GROCERY 68332
MEAT - FRESH & FROZEN 9477
DAIRY 43820
PETS AND SUPPLIES 9761
HOUSEHOLD CHEMICALS/SUPP 24880
nan 0
IMPULSE MERCHANDISE 28712
PRODUCE 51115
CANDY, TOBACCO, COOKIES 9808
GROCERY DRY GOODS 70402
BOYS WEAR 3892
FABRICS AND CRAFTS 5051
JEWELRY AND SUNGLASSES 2526
MENS WEAR 12588
ACCESSORIES 1318
HOME MANAGEMENT 6349
FROZEN FOODS 21101
SERVICE DELI 10127
INFANT CONSUMABLE HARDLINES 10646
PRE PACKED DELI 9052
COOK AND DINE 7600
PHARMACY OTC 23306
LADIESWEAR 8721
COMM BREAD 15244
BAKERY 7168
HOUSEHOLD PAPER GOODS 16274
CELEBRATION 8698
HARDWARE 4913
BEAUTY 14865
AUTOMOTIVE 5414
BOOKS AND MAGAZINES 992
SEAFOOD 1923
OFFICE SUPPLIES 6578
LAWN AND GARDEN 5269
SHEER HOSIERY 434
WIRELESS 1898
BEDDING 2307
BATH AND SHOWER 4582
HORTICULTURE AND ACCESS 2893
HOME DECOR 4167
TOYS 6107
INFANT APPAREL 6051
LADIES SOCKS 775
PLUS AND MATERNITY 591
ELECTRONICS 3190
GIRLS WEAR, 4-6X  AND 7-14 4516
BRAS & SHAPEW

In [26]:
testing_df = train_df['FinelineNumber'].value_counts().reset_index()
testing_df.index = testing_df['index']
testing_df.drop(columns='index',inplace=True)
testing_df['ratio'] = testing_df['FinelineNumber'] * 100 / len(train_df['FinelineNumber'])
testing_df

FinelineNumber     ratio
index                           
5501.0            8244  1.274082
1508.0            5121  0.791433
135.0             4538  0.701333
808.0             4401  0.680160
0.0               3837  0.592995
9546.0            2997  0.463176
1407.0            2848  0.440149
4606.0            2753  0.425467
115.0             2723  0.420830
203.0             2639  0.407848
100.0             2634  0.407076
3004.0            2607  0.402903
4010.0            2531  0.391157
3601.0            2465  0.380957
3600.0            2455  0.379412
110.0             2237  0.345721
3120.0            2074  0.320530
3555.0            2056  0.317748
8101.0            2037  0.314811
5017.0            1980  0.306002
801.0             1946  0.300748
7010.0            1935  0.299048
5620.0            1896  0.293020
202.0             1895  0.292866
9100.0            1885  0.291320
4624.0            1834  0.283438
3467.0            1751  0.270611
7955.0            1749  0.270302
9101.0            1721  0.265975
3702.0            1655  0.255775
...                ...       ...
6213.0               1  0.000155
2289.0               1  0.000155
5380.0               1  0.000155
8539.0               1  0.000155
7231.0               1  0.000155
4314.0               1  0.000155
5578.0               1  0.000155
8554.0               1  0.000155
1152.0               1  0.000155
7773.0               1  0.000155
8565.0               1  0.000155
6177.0               1  0.000155
1765.0               1  0.000155
5564.0               1  0.000155
238.0                1  0.000155
5217.0               1  0.000155
5218.0               1  0.000155
3496.0               1  0.000155
1791.0               1  0.000155
3492.0               1  0.000155
7159.0               1  0.000155
5518.0               1  0.000155
2861.0               1  0.000155
6227.0               1  0.000155
4495.0               1  0.000155
2311.0               1  0.000155
524.0                1  0.000155
7130.0               1  0.000155
2847.0               1  0.000155
7313.0               1  0.000155

[5195 rows x 2 columns]

In [162]:
%%time
ls = []
for i in testing_df.index:
    if len(train_df[['DepartmentDescription','FinelineNumber']][train_df['FinelineNumber'] == i]) == 1:
        ls.append(i)
len(ls)

CPU times: user 24.4 s, sys: 0 ns, total: 24.4 s
Wall time: 24.4 s


In [165]:
print('length of FinelineNumber which points only on DepartmentDescription: ',len(ls),end='\n\n\n'),print(ls)

length of FinelineNumber which points only on DepartmentDescription:  241


[5358.0, 2086.0, 4845.0, 1808.0, 6598.0, 8346.0, 5353.0, 6629.0, 8345.0, 2039.0, 9937.0, 4135.0, 9943.0, 4685.0, 4597.0, 1722.0, 6453.0, 3358.0, 3311.0, 6525.0, 6561.0, 392.0, 7320.0, 8088.0, 6909.0, 6514.0, 6615.0, 6618.0, 5437.0, 1746.0, 2249.0, 2388.0, 9932.0, 4049.0, 3408.0, 3367.0, 1739.0, 3365.0, 8185.0, 4574.0, 9924.0, 6304.0, 7236.0, 6623.0, 2714.0, 2713.0, 6913.0, 8352.0, 8113.0, 259.0, 4751.0, 2649.0, 2648.0, 8257.0, 5387.0, 7216.0, 6744.0, 6813.0, 6793.0, 8259.0, 5383.0, 3795.0, 1336.0, 3338.0, 9792.0, 1691.0, 2621.0, 4761.0, 8087.0, 3430.0, 4764.0, 3794.0, 8073.0, 6482.0, 4090.0, 7160.0, 6513.0, 634.0, 1334.0, 6668.0, 6458.0, 4616.0, 6559.0, 2418.0, 5216.0, 6674.0, 3385.0, 6685.0, 2661.0, 4674.0, 3423.0, 2475.0, 9718.0, 3801.0, 9991.0, 3421.0, 45.0, 2328.0, 3673.0, 1235.0, 5668.0, 8992.0, 2266.0, 4341.0, 6140.0, 3648.0, 5710.0, 3916.0, 1140.0, 384.0, 196.0, 3616.0, 1452.0, 6108.0, 1877.0, 830.0, 517

(None, None)

ls 리스트에 담겨져 있는 FinelineNumber 는 단 하나의 DepartmentDescription을 가리키는 고유한 즉, 말 그래도 FinelineNumber이다.

In [170]:
testing_df2 = train_df['Upc'].value_counts().reset_index()
testing_df2.index = testing_df2['index']
testing_df2.drop(columns='index',inplace=True)
testing_df2['ratio'] = testing_df2['Upc'] * 100 / len(train_df['FinelineNumber'])
testing_df2

Upc     ratio
index                       
4.011000e+03  7657  1.183363
6.053886e+10  2501  0.386521
7.874235e+09  2067  0.319448
7.874235e+09  1883  0.291011
4.046000e+03  1652  0.255311
6.811311e+10  1515  0.234138
6.053887e+10  1487  0.229811
3.338320e+09  1405  0.217138
4.087000e+03  1370  0.211729
6.053887e+10  1343  0.207556
4.900000e+09  1018  0.157328
4.062000e+03   973  0.150374
4.065000e+03   953  0.147283
4.900003e+09   896  0.138474
3.338365e+09   856  0.132292
7.874235e+09   828  0.127965
4.900005e+09   815  0.125955
6.811316e+10   778  0.120237
6.053890e+10   778  0.120237
4.078000e+03   750  0.115910
6.992216e+10   745  0.115137
7.874211e+09   719  0.111119
4.093000e+03   698  0.107874
4.900001e+09   697  0.107719
2.096650e+10   686  0.106019
6.053882e+10   644  0.099528
7.874235e+09   622  0.096128
4.051000e+03   620  0.095819
7.225004e+09   618  0.095510
3.151000e+03   561  0.086701
...            ...       ...
7.413001e+09     1  0.000155
8.806156e+11     1  0.000155
7.278511e+09     1  0.000155
7.144151e+10     1  0.000155
6.811311e+10     1  0.000155
7.169143e+09     1  0.000155
7.583857e+10     1  0.000155
7.943761e+10     1  0.000155
7.871544e+09     1  0.000155
2.700042e+09     1  0.000155
7.659406e+10     1  0.000155
7.163396e+07     1  0.000155
7.595452e+10     1  0.000155
7.617070e+10     1  0.000155
8.434540e+10     1  0.000155
7.575587e+10     1  0.000155
4.402124e+09     1  0.000155
8.809378e+11     1  0.000155
8.108210e+10     1  0.000155
7.163405e+07     1  0.000155
8.511845e+09     1  0.000155
6.821022e+10     1  0.000155
2.898194e+09     1  0.000155
1.675132e+09     1  0.000155
8.888232e+10     1  0.000155
8.437471e+10     1  0.000155
7.644011e+09     1  0.000155
1.119231e+09     1  0.000155
7.670528e+10     1  0.000155
8.669413e+09     1  0.000155

[97714 rows x 2 columns]

In [173]:
%%time
ls2 = []
for i in testing_df2.index:
    if len(train_df[['DepartmentDescription','Upc']][train_df['Upc'] == i]) == 1:
        ls2.append(i)
len(ls2)

CPU times: user 7min 54s, sys: 79.4 ms, total: 7min 54s
Wall time: 7min 54s


In [187]:
len(train_df['Upc'].unique())

97715

In [186]:
print('length of Upc which points only on DepartmentDescription: ',len(ls2),end='\n\n\n'),print(ls2[:10])

length of Upc which points only on DepartmentDescription:  37080


[88786030657.0, 1305139482.0, 4269428770.0, 7644086478.0, 2471931796.0, 88614490328.0, 2856272104.0, 88589800008.0, 88711764844.0, 2856277258.0]


(None, None)

In [185]:
testing_df2.index[0]

4011.0

In [184]:
train_df[['DepartmentDescription','Upc']][train_df['Upc'] == 4046.0].groupby('DepartmentDescription').count()

Upc
DepartmentDescription      
PRODUCE                1652

In [27]:
%%time
ls = []
for i in testing_df.index:
    ls.append(len(train_df[['DepartmentDescription','FinelineNumber']][train_df['FinelineNumber'] == i]))
len(ls)

CPU times: user 28.2 s, sys: 0 ns, total: 28.2 s
Wall time: 28.2 s


In [28]:
ls1 = []
for i in testing_df.index:
    ls1.append(i)

In [29]:
dict_ls_ls1 = dict(zip(ls1,ls))

In [30]:
df = pd.DataFrame()
df['FinelineNumber'] = ls1
df['Duplicated_num'] = ls

In [36]:
train_df[['DepartmentDescription','FinelineNumber']][train_df['FinelineNumber']==5501]

DepartmentDescription  FinelineNumber
28                   PRODUCE          5501.0
66                   PRODUCE          5501.0
77                   PRODUCE          5501.0
111                  PRODUCE          5501.0
198                  PRODUCE          5501.0
207                  PRODUCE          5501.0
268                  PRODUCE          5501.0
370                  PRODUCE          5501.0
663                  PRODUCE          5501.0
681                  PRODUCE          5501.0
738                  PRODUCE          5501.0
786                  PRODUCE          5501.0
823                  PRODUCE          5501.0
885                  PRODUCE          5501.0
933                  PRODUCE          5501.0
1110                 PRODUCE          5501.0
1641                 PRODUCE          5501.0
1705                 PRODUCE          5501.0
1881                 PRODUCE          5501.0
1933                 PRODUCE          5501.0
1958                 PRODUCE          5501.0
2120                 PRODUCE          5501.0
2236                 PRODUCE          5501.0
2272                 PRODUCE          5501.0
2439                 PRODUCE          5501.0
2452                 PRODUCE          5501.0
2533                 PRODUCE          5501.0
2575                 PRODUCE          5501.0
2587                 PRODUCE          5501.0
2771                 PRODUCE          5501.0
...                      ...             ...
645408               PRODUCE          5501.0
645465               PRODUCE          5501.0
645474               PRODUCE          5501.0
645534               PRODUCE          5501.0
645566               PRODUCE          5501.0
645601               PRODUCE          5501.0
645681               PRODUCE          5501.0
645717               PRODUCE          5501.0
645819               PRODUCE          5501.0
645832               PRODUCE          5501.0
645853               PRODUCE          5501.0
645928               PRODUCE          5501.0
646022               PRODUCE          5501.0
646086               PRODUCE          5501.0
646251               PRODUCE          5501.0
646345               PRODUCE          5501.0
646393               PRODUCE          5501.0
646420               PRODUCE          5501.0
646468               PRODUCE          5501.0
646479               PRODUCE          5501.0
646486               PRODUCE          5501.0
646497               PRODUCE          5501.0
646584               PRODUCE          5501.0
646631               PRODUCE          5501.0
646635               PRODUCE          5501.0
646644               PRODUCE          5501.0
646817               PRODUCE          5501.0
646851               PRODUCE          5501.0
646911               PRODUCE          5501.0
646941               PRODUCE          5501.0

[8244 rows x 2 columns]

In [32]:
df

FinelineNumber  Duplicated_num
0             5501.0            8244
1             1508.0            5121
2              135.0            4538
3              808.0            4401
4                0.0            3837
5             9546.0            2997
6             1407.0            2848
7             4606.0            2753
8              115.0            2723
9              203.0            2639
10             100.0            2634
11            3004.0            2607
12            4010.0            2531
13            3601.0            2465
14            3600.0            2455
15             110.0            2237
16            3120.0            2074
17            3555.0            2056
18            8101.0            2037
19            5017.0            1980
20             801.0            1946
21            7010.0            1935
22            5620.0            1896
23             202.0            1895
24            9100.0            1885
25            4624.0            1834
26            3467.0            1751
27            7955.0            1749
28            9101.0            1721
29            3702.0            1655
...              ...             ...
5165          6213.0               1
5166          2289.0               1
5167          5380.0               1
5168          8539.0               1
5169          7231.0               1
5170          4314.0               1
5171          5578.0               1
5172          8554.0               1
5173          1152.0               1
5174          7773.0               1
5175          8565.0               1
5176          6177.0               1
5177          1765.0               1
5178          5564.0               1
5179           238.0               1
5180          5217.0               1
5181          5218.0               1
5182          3496.0               1
5183          1791.0               1
5184          3492.0               1
5185          7159.0               1
5186          5518.0               1
5187          2861.0               1
5188          6227.0               1
5189          4495.0               1
5190          2311.0               1
5191           524.0               1
5192          7130.0               1
5193          2847.0               1
5194          7313.0               1

[5195 rows x 2 columns]

In [34]:
df['Duplicated_num'].describe()

count    5195.000000
mean      123.758422
std       298.338825
min         1.000000
25%        10.000000
50%        35.000000
75%       115.000000
max      8244.000000
Name: Duplicated_num, dtype: float64

In [55]:
dfls1 = df[df['Duplicated_num']<10]['FinelineNumber'].values
len(dfls1)

1271

In [56]:
dfls2 = df[df['Duplicated_num']>115]['FinelineNumber'].values
len(dfls2)

1294

In [58]:
dfls = len(list(dfls1) + list(dfls2))

In [62]:
train_df = pd.read_csv('train.csv')
test_df = pd.read_csv('test.csv')

In [65]:
train_df['FinelineNumber']

1508.0

In [60]:
train_df = make_df_we_wanted(train_df,train_df,test_df,is_need_null_column=True)

##### Dummy타입으로 만든 컬럼 명 : DepartmentDescription

##### ScanCount는 양수만 사용

##### Train df 만드는 중

##### Null 컬럼을 만듬

> 위 정보들을 Display하는 이유는 이번 FeatureMatrix를 사용한 모델에 Fit할 Test 모델 만들 때 같은 전처리를 하기 위해서다.